Lib/zipfile.py

```diff
@@ -571,15 +572,53 @@ def decrypter(data):
 
 
 class LZMACompressor:
-
-    def __init__(self):
+    # LZMA uses "filter chains" to control compression settings (e.g., how
+    # much memory and CPU to use) in details. liblzma defines certain filter
+    # chain configurations as "preset levels," akin to the compression levels
+    # in zlib or bzip2.
+    _PresetOptions = collections.namedtuple(
+        '_PresetOptions',
+        ('lc', 'lp', 'pb', 'dict_size', 'mode', 'mf', 'nice_len', 'depth')
+    )
+    _PRESET_OPTIONS_MAP = {
+        # Levels 0 through 9, without the "extreme" setting
+        0x00000000: _PresetOptions(3, 0, 2, 1 << 18, 1, 3, 128, 4),
+        0x00000001: _PresetOptions(3, 0, 2, 1 << 20, 1, 4, 128, 8),
+        0x00000002: _PresetOptions(3, 0, 2, 1 << 21, 1, 4, 273, 24),
+        0x00000003: _PresetOptions(3, 0, 2, 1 << 22, 1, 4, 273, 48),
+        0x00000004: _PresetOptions(3, 0, 2, 1 << 22, 2, 20, 16, 0),
+        0x00000005: _PresetOptions(3, 0, 2, 1 << 23, 2, 20, 32, 0),
+        0x00000006: _PresetOptions(3, 0, 2, 1 << 23, 2, 20, 64, 0),
+        0x00000007: _PresetOptions(3, 0, 2, 1 << 24, 2, 20, 64, 0),
+        0x00000008: _PresetOptions(3, 0, 2, 1 << 25, 2, 20, 64, 0),
+        0x00000009: _PresetOptions(3, 0, 2, 1 << 26, 2, 20, 64, 0),
+        # Levels 0 through 9, OR-ed with the "extreme" setting (PRESET_EXTREME)
+        0x80000000: _PresetOptions(3, 0, 2, 1 << 18, 2, 20, 273, 512),
+        0x80000001: _PresetOptions(3, 0, 2, 1 << 20, 2, 20, 273, 512),
+        0x80000002: _PresetOptions(3, 0, 2, 1 << 21, 2, 20, 273, 512),
+        0x80000003: _PresetOptions(3, 0, 2, 1 << 22, 2, 20, 192, 0),
+        0x80000004: _PresetOptions(3, 0, 2, 1 << 22, 2, 20, 273, 512),
+        0x80000005: _PresetOptions(3, 0, 2, 1 << 23, 2, 20, 192, 0),
+        0x80000006: _PresetOptions(3, 0, 2, 1 << 23, 2, 20, 273, 512),
+        0x80000007: _PresetOptions(3, 0, 2, 1 << 24, 2, 20, 273, 512),
+        0x80000008: _PresetOptions(3, 0, 2, 1 << 25, 2, 20, 273, 512),
+        0x80000009: _PresetOptions(3, 0, 2, 1 << 26, 2, 20, 273, 512),
+    }
+
+    def __init__(self, preset=None):
         self._comp = None
+        if (preset is not None) and (preset not in self._PRESET_OPTIONS_MAP):
```

Comment:
I don't think the parentheses here are needed.

Reply:
Indeed. My thought was that it helped with readability.